### ISSUE TITLE
After lws_context_destroy, calling curl always got 'Out of Memory' error 

### ISSUE THREAD

**Issue by slend0225** [warmcat/libwebsockets#569]
I have a program that have an websocket for receiving audio data from audio server. I called `lws_create_context` to create an context and use this context to call `lws_client_connect` to construct an `wsiAudio`.

And when my main program got some condition that need to shut down the audio channel, I will call `lws_context_destroy` with the relative `wsiAudio`'s context.

But I found after `lws_context_destroy` have been called, every time my main program call  `curl` will get error code `27 (CURLE_OUT_OF_MEMORY) `or occasionally `35 (CURLE_SSL_CONNECT_ERROR)`.

And if I don't call `lws_context_destroy` in any condition, the curl will work fine.

I use libwebsocket 1.7.8.

**Comment by lws-team** [comment]
What exactly is your "main program" in this?

It's the thing that uses lws, or something else?

While a process is alive, it can hang on to unfreed memory okay.

So if the problem is your 'main program' allocates and never frees a ton of memory, and never exits either (you mention your 'main program' is 'calling curl' 'after calling lws_context_destroy),it can add up.

But I'm not sure if we're talking about that or something else.

**Comment by slend0225** [comment]
Sorry for the wrong and unclear description.
What I mean 'main program' is more likely to indicate 'main process'

I re-describe my issue, hope it let you more clear to it.

My program is a multi-thread program, 
one of the threads do constructing and destroy websocket for audio receiving, 
another thread do constructing and destroy websocket for receiving command messages from message server, 
and the main thread use the command message or audio data, and in some command, it will call `curl`.

In short, I have a program, which have two threads that each construct a wsi with different context (and destroy each context in each thread), and one thread that will call curl in some cases.

The problem is that when one of the websocket thread destroy a context for a wsi, my main thread call curl will get error code` 27 (CURLE_OUT_OF_MEMORY)`or occasionally `35 (CURLE_SSL_CONNECT_ERROR)`.

**Comment by JoakimSoderberg** [comment]
You need to be a lot more specific in showing that this is an LWS issue and not your own code leaking memory.

I'd suggest you run your program through valgrind (if using unix?) or Dr Memory on windows and fix all errors (note OpenSSL will create false positives accessing random memory though).

Most likely this is a memory leak, not caused by LWS but your program. But having run the above you will have more relevant information if it is an LWS issue.

**Comment by lws-team** [comment]
As Joakim says it's a bit all mixed together, the only thing to go on is the problem comes after context destroy.  Actually this multi context thing is not tested, lws is designed to do everything in one context / event loop (btw the client stuff in lws can also do wget / curl http[s] client tasks alongside ws, interleaved without threads).

I took a look at context destroy and as you would expect everything is properly related to cleaning stuff in the specific context, lws references and allocates everything of its own strictly inside the context.  The only funny business I can see is this that ends up getting called:

```
void
lws_ssl_context_destroy(struct lws_context *context)
{
#if defined(LWS_USE_POLARSSL)
#else
#if defined(LWS_USE_MBEDTLS)
#else
#if (OPENSSL_VERSION_NUMBER < 0x01000000) || defined(USE_WOLFSSL)
	ERR_remove_state(0);
#else
#if (OPENSSL_VERSION_NUMBER >= 0x10100005L) && \
	!defined(LIBRESSL_VERSION_NUMBER) && \
	!defined(OPENSSL_IS_BORINGSSL)
	ERR_remove_thread_state();
#else
	ERR_remove_thread_state(NULL);
#endif
#endif
	ERR_free_strings();
	EVP_cleanup();
	CRYPTO_cleanup_all_ex_data();
#endif
#endif
}
```

Maybe when one thread does (the working) openssl cleanup, some of it affects the whole process ability to use openssl.

If you're running openssl <1.0, you can see that alone did not have a perthread destroy api itself and would likely trash ability of other threads to continue to use it.   Things like CRYPTO_cleanup_all_ext_data(), an openssl api, sound processwide.

Either way my guess is the threads might "communicate by openssl state" unintentionally.  You might try comment the call to lws_ssl_context_destroy() in ./lib/context.c, accepting we will leak there on process exit, as a test and see if the symptom is affected.

**Comment by slend0225** [comment]
Thank you for your replying.
Because this code is actually run on a embedded system, I can't easily apply those memory check thing on that.

However I tried to comment the `lws_ssl_context_destroy()` and re-build the libwebsocket lib, and wuth this 'revised' libwebsocket.so.7, somehow the curl thing don't get those error code.

Then I tried which part in `lws_ssl_context_destroy()` may related on this curl error condition, and I found when comment `EVP_cleanup()`, the curl thing can normally return.

I found some information in OpenSSL issue forum about using `EVP_cleanup()`. (see the url below)[https://rt.openssl.org/Ticket/Display.html?id=2217&user=guest&pass=guest](url)

It seems `EVP_cleanup()` will affect other modules that also use OpenSSL. 
And It seems having some solution recently with auto-init/-deinit and thread-once facilities.

**Comment by lws-team** [comment]
Hum I am not sure what you mean by 'module'.  Openssl is a library and these bind to the *process*, they stick the sections from the module into the user process address space.

Threads are subordinate to the process, so all the threads in the process see the same thing.  We don't need to introduce 'modules' whatever they are.

The 'solution' on that bug (after 6 years...) is just on openssl master since March, almost everyone is getting much older versions from their distro.

Actually the normal flow for lws is you destroy your one lws context when your process exits, then there is no problem.

How about you look at ideally just using one context from one thread, or if you can't, at least let the context lifetimes match the process lifetimes.  Then this gnarly problem will just disappear.

**Comment by lws-team** [comment]
I think openssl 1.0.2 and below that is widely used doesn't really have a way to solve this.  As another user pointed out with a similar question a few months ago, we can't know either that lws is the only library bound to the process that uses openssl, so our lws-only reference counting for processwide openssl destroy won't actually solve it.  In your case libcurl will init and destroy openssl according to its own ideas.

The only way to solve it atm I think is use one lws context and destroy it at process exit, if libcurl destroys the processwide openssl pieces too early look at doing client operations over https also in lws using the one context, this works in master.
